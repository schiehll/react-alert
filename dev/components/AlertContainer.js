import React from 'react';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
import AlertMessage from './AlertMessage';
import reactAlertEvents from '../reactAlertEvents';

class AlertContainer extends React.Component {

  constructor(props){
    super(props);
    this.state = {
      alerts: []
    };
    this.style = this._setStyle();
    this.theme = this._setTheme();
    this._eventListners();
  }
  /**
   * Show the alert in the page with success type
   * @param  {string} message
   * @param  {Object} options
   * @return {void}
   */
  success(message, options = {}){
    options.type = 'success';
    this.show(message, options);
  }
  /**
   * Show the alert in the page with error type
   * @param  {string} message
   * @param  {Object} options
   * @return {void}
   */
  error(message, options = {}){
    options.type = 'error';
    this.show(message, options);
  }
  /**
   * Show the alert in the page with info type
   * @param  {string} message
   * @param  {Object} options
   * @return {void}
   */
  info(message, options = {}){
    options.type = 'info';
    this.show(message, options);
  }
  /**
   * Show the alert in the page
   * @param  {string} message
   * @param  {Object} options
   * @return {void}
   */
  show(message, options = {}){
    let alert = {};
    alert.message = message;
    alert = Object.assign(alert, options);
    this.setState({alerts: this._addAlert(alert)});
  }
  /**
   * Remove all tasks from the page
   * @return {void}
   */
  removeAll(){
    this.setState({alerts: []});
  }
  /**
   * Add an alert
   * @param {Object} alert
   */
  _addAlert(alert){
    alert.uniqueKey = this._genUniqueKey();
    alert.style = this.theme;
    if(!alert.hasOwnProperty('time')){
      alert.time = this.props.time;
    };
    alert.closeIconClass = 'close-' + this.props.theme;
    this.state.alerts.push(alert);
    return this.state.alerts;
  }
  /**
   * Generate a key
   * @return {string}
   */
  _genUniqueKey(){
    return new Date().getTime().toString() + Math.random().toString(36).substr(2, 5);
  }
  /**
   * Remove an AlertMessage from the container
   * @param  {AlertMessage} alert
   * @return {void}
   */
  _removeAlert(alert){
    return this.state.alerts.filter((a) => {
      return a.uniqueKey != alert.props.uniqueKey;
    });
  }
  /**
   * Listen to alert events
   * @return {void}
   */
  _eventListners(){
    reactAlertEvents.on('ALERT.REMOVE', (alert) => {
      this.setState({alerts: this._removeAlert(alert)});
    });
  }
  /**
   * Set the alert position on the page
   */
  _setStyle(){
    let position = {};
    switch(this.props.position){
      case 'top left':
        position = {
          top: 0,
          right: 'auto',
          bottom: 'auto',
          left: 0
        }
        break;
      case 'top right':
        position = {
          top: 0,
          right: 0,
          bottom: 'auto',
          left: 'auto'
        }
        break;
      case 'bottom left':
        position = {
          top: 'auto',
          right: 'auto',
          bottom: 0,
          left: 0
        }
        break;
      default:
        position = {
          top: 'auto',
          right: 0,
          bottom: 0,
          left: 'auto'
        }
        break;
    }

    return {
      margin: this.props.offset + 'px',
      top: position.top,
      right: position.right,
      bottom: position.bottom,
      left: position.left,
    };
  }
  /**
   * Set the style of the alert based on the chosen theme
   */
  _setTheme(){
    let theme = {};
    switch(this.props.theme){
      case 'light':
        theme = {
          alert: {
            backgroundColor: '#fff',
            color: '#333'
          },
          closeButton: {
            bg: '#f3f3f3'
          }
        }
        break;
      default:
        theme = {
          alert: {
            backgroundColor: '#333',
            color: '#fff'
          },
          closeButton: {
            bg: '#444'
          }
        }
        break;
    }

    return theme;
  }

  componentDidUpdate(){
    this.style = this._setStyle();
    this.theme = this._setTheme();
  }

  render(){
    return(
      <div style={this.style} className="react-alerts">
        <ReactCSSTransitionGroup
          transitionName={this.props.transition}
          transitionEnterTimeout={250}
          transitionLeaveTimeout={250}>
          {this.state.alerts.map((alert, index) => {
            return <AlertMessage key={alert.uniqueKey} {...alert} />
          })}
        </ReactCSSTransitionGroup>
      </div>
    );
  }
}

AlertContainer.defaultProps = {
  offset: 14,
  position: 'bottom left',
  theme: 'dark',
  time: 5000,
  transition: 'scale'
}

AlertContainer.propTypes = {
  offset: React.PropTypes.number,
  position: React.PropTypes.oneOf([
    'bottom left',
    'bottom right',
    'top right',
    'top left',
  ]),
  theme: React.PropTypes.oneOf(['dark', 'light']),
  time: React.PropTypes.number,
  transition: React.PropTypes.oneOf(['scale', 'fade'])
}

export default AlertContainer;
