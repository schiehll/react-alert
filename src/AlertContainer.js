class AlertContainer {
  constructor(defaultOptions) {
    this.defaultOptions = defaultOptions
    this.listners = []
    this.alerts = []
  }

  show = (message = '', options = {}) => {
    const id = new Date().getTime()
    const alertOptions = {
      ...this.defaultOptions,
      ...options
    }

    const alert = {
      id,
      message,
      options: alertOptions
    }

    alert.close = () => this.remove(alert)

    if (alert.options.timeout) {
      setTimeout(() => {
        this.remove(alert)
      }, alert.options.timeout)
    }

    this.alerts.push(alert)
    this._broadcastToFirstListner()

    alert.options.onOpen && alert.options.onOpen()

    return alert
  }

  remove = alert => {
    const lengthBeforeRemove = this.alerts.length
    this.alerts = this.alerts.filter(a => a.id !== alert.id)

    if (lengthBeforeRemove > this.alerts.length) {
      this._broadcastToFirstListner()
      alert.options.onClose && alert.options.onClose()
    }
  }

  success = (message = '', options = {}) => {
    options.type = 'success'
    return this.show(message, options)
  }

  error = (message = '', options = {}) => {
    options.type = 'error'
    return this.show(message, options)
  }

  info = (message = '', options = {}) => {
    options.type = 'info'
    return this.show(message, options)
  }

  _getAlerts = () => this.alerts

  _broadcastToFirstListner = () => {
    this.listners.length && this.listners[0](this.alerts)
  }

  _subscribe = listener => {
    this.listners.push(listener)
  }

  _unsubscribe = listener => {
    this.listners = this.listners.filter(item => item !== listener)
  }
}

export default AlertContainer
