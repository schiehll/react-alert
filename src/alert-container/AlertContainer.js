import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {CSSTransitionGroup} from 'react-transition-group'
import shortid from 'shortid'
import Container from './Container'
import AlertMessage from '../alert'

class AlertContainer extends Component {
  static defaultProps = {
    offset: 14,
    position: 'bottom left',
    theme: 'dark',
    time: 5000,
    transition: 'scale'
  }

  static propTypes = {
    offset: PropTypes.number,
    position: PropTypes.oneOf([
      'bottom left',
      'bottom right',
      'top right',
      'top left'
    ]),
    theme: PropTypes.oneOf(['dark', 'light']),
    time: PropTypes.number,
    transition: PropTypes.oneOf(['scale', 'fade'])
  }

  state = {
    alerts: []
  }

  success = (message = '', options = {}) => {
    options.type = 'success'
    return this.show(message, options)
  }

  error = (message = '', options = {}) => {
    options.type = 'error'
    return this.show(message, options)
  }

  info = (message = '', options = {}) => {
    options.type = 'info'
    return this.show(message, options)
  }

  show = (message = '', options = {}) => {
    const {theme, time} = this.props

    const alert = {
      id: shortid.generate(),
      message,
      time,
      theme,
      ...options
    }

    this.setState(prevState => ({
      alerts: prevState.alerts.concat(alert)
    }))
    return alert.id;
  }

  removeAll = () => {
    const alertsRemoved = this.state.alerts
    this.setState({alerts: []})
    alertsRemoved.forEach(alert => alert.onClose && alert.onClose())
  }

  removeAlert = id => {
    const alertRemoved = this.state.alerts.filter(alert => alert.id === id)[0]
    this.setState(prevState => ({
      alerts: prevState.alerts.filter(alert => alert.id !== id)
    }))
    alertRemoved && alertRemoved.onClose && alertRemoved.onClose()
  }

  render () {
    const {position, offset, transition} = this.props

    return (
      <Container glam={{position, offset}}>
        <CSSTransitionGroup
          transitionName={transition}
          transitionEnterTimeout={250}
          transitionLeaveTimeout={250}
        >
          {this.state.alerts.map(alert => {
            return (
              <AlertMessage
                key={alert.id}
                {...alert}
                onRemoveAlert={this.removeAlert}
              />
            )
          })}
        </CSSTransitionGroup>
      </Container>
    )
  }
}

export default AlertContainer
