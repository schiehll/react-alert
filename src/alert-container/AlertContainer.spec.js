import React from 'react'
import renderer from 'react-test-renderer'
import {shallow} from 'enzyme'
import AlertContainer from './AlertContainer'

describe('AlertContainer', () => {
  test('should set its position based on the given position prop', () => {
    let props = {
      position: 'top left'
    }
    let component = renderer.create(
      <AlertContainer {...props} />
    )

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    props = {
      position: 'top right'
    }
    component = renderer.create(
      <AlertContainer {...props} />
    )

    tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    props = {
      position: 'bottom right'
    }
    component = renderer.create(
      <AlertContainer {...props} />
    )

    tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    props = {
      position: 'bottom left'
    }
    component = renderer.create(
      <AlertContainer {...props} />
    )

    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('show', () => {
    test('should be called on each type method call', () => {
      const wrapper = shallow(<AlertContainer />)
      const instance = wrapper.instance()
      instance.show = jest.fn()

      instance.success('Some message')
      instance.info('Some message')
      instance.error('Some message')

      expect(instance.show).toHaveBeenCalledTimes(3)
      expect(instance.show.mock.calls[0][0]).toBe('Some message')
      expect(instance.show.mock.calls[0][1]).toEqual({ type: 'success' })
      expect(instance.show.mock.calls[1][0]).toBe('Some message')
      expect(instance.show.mock.calls[1][1]).toEqual({ type: 'info' })
      expect(instance.show.mock.calls[2][0]).toBe('Some message')
      expect(instance.show.mock.calls[2][1]).toEqual({ type: 'error' })
    })

    test('should add an alert object to state', () => {
      const wrapper = shallow(<AlertContainer />)
      const instance = wrapper.instance()

      expect(instance.state.alerts).toHaveLength(0)
      instance.show('Some message')
      expect(instance.state.alerts).toHaveLength(1)
    })

    test('should be called with default method params', () => {
      const wrapper = shallow(<AlertContainer />)
      const instance = wrapper.instance()
      instance.show = jest.fn()

      instance.success()
      instance.info()
      instance.error()

      expect(instance.show).toHaveBeenCalledTimes(3)
      expect(instance.show.mock.calls[0][0]).toBe('')
      expect(instance.show.mock.calls[0][1]).toEqual({ type: 'success' })
      expect(instance.show.mock.calls[1][0]).toBe('')
      expect(instance.show.mock.calls[1][1]).toEqual({ type: 'info' })
      expect(instance.show.mock.calls[2][0]).toBe('')
      expect(instance.show.mock.calls[2][1]).toEqual({ type: 'error' })
    })

    test('should work with default options', () => {
      const wrapper = shallow(<AlertContainer />)
      const instance = wrapper.instance()

      instance.show()
      expect(instance.state.alerts[0].id).toBeDefined()
      expect(instance.state.alerts[0].message).toBe('')
      expect(instance.state.alerts[0].time).toBe(5000)
      expect(instance.state.alerts[0].theme).toBe('dark')
    })

    test('should work with custom options', () => {
      const wrapper = shallow(<AlertContainer time={123} theme='light' />)
      const instance = wrapper.instance()

      instance.show('custom message', { foo: 1, bar: 2 })
      expect(instance.state.alerts[0].id).toBeDefined()
      expect(instance.state.alerts[0].message).toBe('custom message')
      expect(instance.state.alerts[0].time).toBe(123)
      expect(instance.state.alerts[0].theme).toBe('light')
      expect(instance.state.alerts[0].foo).toBe(1)
      expect(instance.state.alerts[0].bar).toBe(2)
    })
  })

  describe('removeAll', () => {
    test('should remove all alerts from state, calling onClose function', () => {
      const wrapper = shallow(<AlertContainer />)
      const instance = wrapper.instance()
      const onCloseFn1 = jest.fn()
      const onCloseFn2 = jest.fn()

      expect(instance.state.alerts).toHaveLength(0)
      instance.show('Some message', { onClose: onCloseFn1 })
      instance.show('Some message', { onClose: onCloseFn2 })
      expect(instance.state.alerts).toHaveLength(2)
      instance.removeAll()
      expect(instance.state.alerts).toHaveLength(0)
      expect(onCloseFn1).toHaveBeenCalled()
      expect(onCloseFn2).toHaveBeenCalled()
    })
  })

  describe('removeAlert', () => {
    test('should remove a given alert from state', () => {
      const wrapper = shallow(<AlertContainer />)
      const instance = wrapper.instance()
      const onCloseFn = jest.fn()

      expect(instance.state.alerts).toHaveLength(0)
      instance.show('Some message', { onClose: onCloseFn })
      expect(instance.state.alerts).toHaveLength(1)
      instance.removeAlert(instance.state.alerts[0].id)
      expect(instance.state.alerts).toHaveLength(0)
      expect(onCloseFn).toHaveBeenCalled()
    })

    test('should not call close function if alert is manually closed', () => {
      const wrapper = shallow(<AlertContainer />)
      const instance = wrapper.instance()
      const onCloseFn = jest.fn()

      expect(instance.state.alerts).toHaveLength(0)
      instance.show('Some message', { onClose: onCloseFn })
      expect(instance.state.alerts).toHaveLength(1)
      instance.removeAlert(instance.state.alerts[0].id)

      onCloseFn.mockReset()
      
      instance.removeAlert('deleted id')
      expect(instance.state.alerts).toHaveLength(0)
      expect(onCloseFn).not.toHaveBeenCalled()
    })
  })
})
