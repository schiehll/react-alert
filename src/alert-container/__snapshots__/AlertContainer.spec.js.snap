// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertContainer should set its position based on the given position prop 1`] = `
<div
  className="css-9xlrwr"
>
  <span />
</div>
`;

exports[`AlertContainer should set its position based on the given position prop 2`] = `
<div
  className="css-132hx50"
>
  <span />
</div>
`;

exports[`AlertContainer should set its position based on the given position prop 3`] = `
<div
  className="css-15ol66j"
>
  <span />
</div>
`;

exports[`AlertContainer should set its position based on the given position prop 4`] = `
<div
  className="css-1bdcmy7"
>
  <span />
</div>
`;

exports[`AlertContainer should set its position based on the given position prop 5`] = `
<div
  className="css-1e2jhbr"
>
  <span />
</div>
`;

exports[`AlertContainer should set its position based on the given position prop 6`] = `
<div
  className="css-r7ps70"
>
  <span />
</div>
`;
